timeseries/src/autogluon/timeseries/models/ensemble/abstract.py

```diff
@@ -0,0 +1,136 @@
+import functools
+import logging
+from abc import ABC
+from typing import Dict, List, Optional
+
+import numpy as np
+from typing_extensions import final
+
+from autogluon.core.utils.exceptions import TimeLimitExceeded
+from autogluon.timeseries.dataset import TimeSeriesDataFrame
+from autogluon.timeseries.models.abstract import TimeSeriesModelBase
+
+logger = logging.getLogger(__name__)
+
+
+class AbstractTimeSeriesEnsembleModel(TimeSeriesModelBase, ABC):
+    """Abstract class for time series ensemble models."""
+
+    @property
+    def model_names(self) -> List[str]:
+        """Names of base models included in the ensemble."""
+        raise NotImplementedError
+
+    @final
+    def fit(
+        self,
+        predictions_per_window: Dict[str, List[TimeSeriesDataFrame]],
+        data_per_window: List[TimeSeriesDataFrame],
+        model_scores: Optional[Dict[str, float]] = None,
+        time_limit: Optional[float] = None,
+    ):
+        """Fit ensemble model given predictions of candidate base models and the true data.
+
+        Parameters
+        ----------
+        predictions_per_window : Dict[str, List[TimeSeriesDataFrame]]
+            Dictionary that maps the names of component models to their respective predictions for each validation
+            window.
+        data_per_window : List[TimeSeriesDataFrame]
+            Observed ground truth data used to train the ensemble for each validation window. Each entry in the list
+            includes both the forecast horizon (for which the predictions are given in ``predictions``), as well as the
+            "history".
+        model_scores : Optional[Dict[str, float]]
+            Scores (higher is better) for the models that will constitute the ensemble.
+        time_limit : Optional[int]
```

Comment:
Nit: `Optional[float]` instead of `Optional[int]`.

In general, I think we can safely skip the type annotations in docstrings for internal classes that are not shown in the API section on https://auto.gluon.ai. The type hints are already available in the method signature, and duplicating them in the docstrings just creates room for typos.

Reply:
I know we discussed this before. Can you maybe open an issue for this? I am not addressing this or other typing issues on the fly in order to address them all in one go for consistency.

Reply:
Sure, makes sense #5074